Upgrade Scale Testing

sarahbx opened the pull request:
- create vms w/ load

Part of a dependency path for scale upgrade testing
Rebased on #3108 

Depends on:
#3149 - Alter chaos multiprocessing start method
#3029 - Move Chaos utils
#3103 - Move EUS Constant
#3105 - Add scale utils.py
#3107 - Add scale conftest.py
#3108 - Add scale threaded utils

##### Short description:

##### More details:

##### What this PR does / why we need it:

##### Which issue(s) this PR fixes:

##### Special notes for reviewer:

##### jira-ticket:
<!--  full-ticket-url needs to be provided. This would add a link to the pull request to the jira and close it when the pull request is merged
If the task is not tracked by a Jira ticket, just write "NONE".
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added end-to-end upgrade-scale test suite and threaded scale utilities for parallel VM orchestration, guest-data collection/validation, and stress-ng workload support.

* **Dependencies**
  * Added runtime dependency: openshift-python-scale-utilities (>=0.1.1.5).

* **Tests**
  * New pytest marker and --scale option, numerous scale/chaos fixtures, multiprocessing start-method fixture, and many new/updated scale test modules.

* **Chores**
  * Updated ignore list to exclude scale test artifacts; added new constants and test utilities to support scale workflows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sarahbx committed:
Move chaos utils

Move chaos utils for use in other modules
- create_vm_with_nginx_service
- verify_vm_service_reachable
- is_http_ok

vsibirsk commented on tests/scale/threaded_utils/virt.py:
```diff
@@ -0,0 +1,184 @@
+    assert vms, "No VMs provided {vms!r}"
+
+    def _wait_for_accessible_vm(_vm: VirtualMachineForTests) -> None:
+        wait_for_ssh_connectivity(vm=_vm, timeout=timeout, tcp_timeout=tcp_timeout, sleep=sleep)
```
pass `wait_for_ssh_connectivity`
no need for inner func here

sarahbx replied:
This inner func is utilized to simplify the the call to `executor.map()` below.
`map()` would require iterables generated for the other arguments

vsibirsk commented on tests/scale/threaded_utils/virt.py:
```diff
@@ -0,0 +1,184 @@
+            host=_vm.ssh_exec,
+            commands=commands,
+            tcp_timeout=tcp_timeout,
+        )
```
ditto

sarahbx replied:
ditto

sarahbx committed:
Add typing, rename args

sarahbx commented on tests/scale/threaded_utils/virt.py:
```diff
@@ -0,0 +1,184 @@
+        except TimeoutExpiredError:
+            LOGGER.error(f"VM: {_vm.name} Status: {_vm.instance.status} virt-launcher: {sample}")
+
+            raise
```
Resource.exists returns either ResourceInstance or None.
There is also a check before accessing via `if sample and sample.spec.nodeName`

sarahbx commented on tests/scale/threaded_utils/virt.py:
```diff
@@ -0,0 +1,184 @@
+    return all_guest_data
+
+
+def verify_guest_data(input_tuple: tuple) -> None:
```
This was another attempt at simplification due to the way executor.map() works. will look into

hmeir reviewed (commented):
@sarahbx Hi :)
I

rnetser reviewed (changes requested):
this is a huge pr; please split to smaller ones. thank yo

sarahbx committed:
Move EUS constant to fix import error

sarahbx committed:
Add scale utils.py

sarahbx committed:
Add scale/conftest.py

sarahbx committed:
Add scale/threaded_utils

sarahbx committed:
Upgrade Scale Testing

- create VMs with load

sarahbx committed:
Add typing, make name required

rnetser commented on tests/scale/test_upgrade_scale.py:
```diff
@@ -0,0 +1,443 @@
+    for entry in request.param:
+        vm_instancetype = VirtualMachineClusterInstancetype(name=entry["vm_instancetype"])
+        vm_preference = VirtualMachineClusterPreference(name=entry["vm_preference"])
+        cloud_init_data = None
```
ditto

sarahbx replied:
done

rnetser commented on tests/scale/test_upgrade_scale.py:
```diff
@@ -0,0 +1,443 @@
+    idle_monitored_api_requests,
+):
+    guest_data_list = threaded_get_vm_guest_data(vms=running_vms_with_load, commands=GUEST_DATA_COMMANDS)
+    request.config.cache.set(f"{cache_key_scope_func}::guest_data_list", guest_data_list)
```
why is this needed?

sarahbx committed:
Alter chaos multiprocessing start method

The previous fix of using get_context(method) is no longer working.
To remedy this, switching to a fixture that sets the
multiprocessing start method instead.
https://github.com/python/cpython/issues/132898
https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process

sarahbx committed:
Add scale pytest argument, filter out scale otherwise

sarahbx committed:
Add missing constants and defaults

sarahbx committed:
Add docstring to get_user_kubeconfig_context

sarahbx committed:
Add MachineConfigPoolConfiguration ExitStack

sarahbx committed:
Refactor to use MachineConfigPoolConfiguration

sarahbx committed:
Update arguments and docstrings, include constants, support admin_client

sarahbx commented on tests/scale/test_upgrade_scale.py:
```diff
@@ -0,0 +1,443 @@
+    pytest.mark.upgrade,
+    pytest.mark.ocp_upgrade,
+    pytest.mark.cnv_upgrade,
+    pytest.mark.eus_upgrade,
```
I'm working on a modification to allow this usage without disrupting general upgrade tests

sarahbx committed:
Add client, split vms_for_upgrade_test

sarahbx commented on tests/scale/test_upgrade_scale.py:
```diff
@@ -0,0 +1,443 @@
+            commands=shlex.split(
+                f"curl -LO {stress_ng_url_for_cirros} && chmod 0755 stress-ng && ./stress-ng --version"
+            ),
+        )
```
done

sarahbx committed:
Correct errors made in refactor

sarahbx committed:
Add admin_client to LocalThreadedScaleResources

sarahbx committed:
Update unittests for operator.py

sarahbx committed:
Add additional unittetst assertion

sarahbx committed:
Fix docstring typo

sarahbx committed:
Move upgrade marks to upgrade test class

sarahbx commented on conftest.py:
```diff
@@ -505,6 +511,14 @@ def filter_sno_only_tests(items: list[Item], config: Config) -> list[Item]:
     return items
 
 
+def filter_scale_only_tests(items: list[Item], config: Config) -> list[Item]:
```
np, we can use the marker. the logic would be:
```
marker_config = config.getoption("-m")
if marker_config and ("scale" not in marker_config or "not scale" in marker_config):
    # remove scale tests
```
It's not an absolute solution for all possible boolean logic, but should cover most cases

openshift-virtualization-qe-bot-3 commented:
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #3789._

<details>
<summary>Overlapping files</summary>

conftest.py
pytest.ini
utilities/constants.py

</details>

openshift-virtualization-qe-bot-3 commented:
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #3901._

<details>
<summary>Overlapping files</summary>

.flake8
conftest.py
pyproject.toml
uv.lock

</details>
